test/agentchat/test_groupchat.py

```diff
@@ -49,6 +51,181 @@ def test_func_call_groupchat():
     agent2.initiate_chat(group_chat_manager, message={"function_call": {"name": "func", "arguments": '{"x": 1}'}})
 
 
+def test_group_chat_math_class():
+    """
+    This test case is to simulate a math class.
+    where teacher creates math questions and student resolves the questions.
+    teacher will create a question, student will resolve the question and tell teacher the answer.
+    If the answer is correct, teacher will create another question, otherwise, teacher will ask student to resolve the question again.
+    The class will end when teacher has created 3 questions.
+
+    This test case is created to test the following features:
+    - speaker selection should work under a continuous q&a scenario among two agents and GPT 3.5 model.
+    - admin should end the class when teacher has created 3 questions.
+    """
+    skip_if_openai_not_available()
+    config_list = autogen.config_list_from_json(
+        OAI_CONFIG_LIST,
+        file_location=KEY_LOC,
+        filter_dict={
+            "model": ["gpt-3.5-turbo"],
+        },
+    )
+    gpt3_5_config = {
+        "model": "gpt-3.5-turbo",
+        "seed": random.randint(0, 100),  # change the seed for different trials
+        "temperature": 0,
+        "config_list": config_list,
+        "request_timeout": 120,
+    }
+
+    llm_config_for_user_proxy = {
+        **gpt3_5_config,
+        "functions": [
+            {
+                "name": "terminate_group_chat",
+                "description": "terminate group chat",
+                "parameters": {
+                    "type": "object",
+                    "properties": {
+                        "message": {
+                            "type": "string",
+                            "description": "terminate group chat message",
+                        },
+                    },
+                    "required": ["message"],
+                },
+            }
+        ],
+    }
+
+    def terminate_group_chat(message):
```

Comment:
@afourney This is how to achieve a more robust terminating strategy via function_call, could you review it?

Reply:
@LittleLittleCloud I am having trouble understand why this strategy is more robust?

This strategy: one of the agents calls a group chat termination function. user proxy executes. manager detects termination string

Old strategy: one of the agent generates a termination string. manager detects.

Reply:
> If an agent is smart enough for suggested strategy then it should be able to also do old strategy?

Yes! Unfortunately, that's not the case in the real world, especially for `gpt-3.5-turbo`. Even for `gpt-4` agents, they might also fail to give the correct termination world when the conversation grows.

The new strategy (using termination_function_call) can make sure the group chat terminate correctly when that termination function_call get triggered. That strategy also works well on gpt-3.5-turbo which fined-tuned for function_call.

Reply:
Got it -- the fact that function calls are prioritized makes it more robust.